source/kernel/src/drivers/emb_display.rs

```diff
@@ -0,0 +1,260 @@
+pub struct FrameChunk {
+    frame_id: u16,
+    bytes: HeapArray<u8>,
+    start_x: i32,
+    start_y: i32,
+    width: u32,
+    height: u32,
+}
+
+struct FrameInfo {
+    frame: u16,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum FrameError {
+    DuplicateItem,
+    RegistryFull,
+}
+
+struct DisplayInfo{
+    kernel: &'static Kernel,
+    frames: HeapFixedVec<FrameInfo>,
+    frame_size: usize,
+}
+
+// Client interface to EmbDisplay
+pub struct EmbDisplayHandle {
+    prod: KernelHandle<EmbDisplay>,
+    reply: Reusable<Envelope<Result<Response, FrameError>>>,
+}
+
+pub enum Request {
+    NewFrameChunk { frame_id: u16, start_x: i32, start_y: i32, width: u32, height: u32 },
+}
+
+pub enum Response {
+    FrameChunkAllocated(FrameChunk),
+}
+
+pub enum EmbDisplayError {
+    BufferFull,
+}
+
+struct CommanderTask {
+    cmd: KConsumer<Message<EmbDisplay>>,
+    fmutex: HeapArc<Mutex<DisplayInfo>>, 
+}
+
+// impl EmbDisplay
+impl RegisteredDriver for EmbDisplay {
+    type Request = Request;
+    type Response = Response;
+    type Error = FrameError;
+    const UUID: Uuid = crate::registry::known_uuids::kernel::EMB_DISPLAY;
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum RegistrationError {
+    DisplayAlreadyExists,
+}
+
+// Register the driver instance
+impl EmbDisplay {
+    pub async fn register(
+        kernel: &'static Kernel,
+        max_frames: usize,
+        width: usize,
+        height: usize,
+    ) -> Result<(), ()> {
+        let frames = kernel.heap().allocate_fixed_vec(max_frames).await;
+        let frame_size = width * height * BYTES_PER_PIXEL as usize;
+
+        let imutex = kernel
+            .heap()
+            .allocate_arc(Mutex::new(DisplayInfo { 
+                kernel,
+                frames,
+                frame_size,
+            }))
+            .await;
+        let (cmd_prod, cmd_cons) = KChannel::new_async(kernel, 1).await.split();
+        let commander = CommanderTask {
+            cmd: cmd_cons,
+            fmutex: imutex,
+        };
+
+        kernel.spawn(async move{
+            commander.run().await;
+        }).await;
+
+        kernel
+            .with_registry(|reg| reg.register_konly::<EmbDisplay>(&cmd_prod))
+            .await
+            .map_err(|_| RegistrationError::DisplayAlreadyExists).unwrap();
+
+        Ok(())
+    }
+}
+
+// Create and return a Simulator display object from raw pixel data.
+// Pixel data is turned into a raw image, and then drawn onto a SimulatorDisplay object
+// This is necessary as a e-g Window only accepts SimulatorDisplay object
+// On a physical display, the raw pixel data can be sent over to the display directly 
+// Using the display's device interface
+impl FrameChunk {
+    pub async fn frame_display(&mut self) -> Result<ImageRaw<Gray8>, ()> {
+        let raw_image: ImageRaw<Gray8>;
+        raw_image = ImageRaw::<Gray8>::new(self.bytes.as_ref(), self.width);
+        Ok(raw_image)
+    }
+
+    pub async fn frame_clear(&mut self) {
```

Comment:
as far as I can tell, neither of these methods _need_ to be async --- they don't `.await` anything. can we change them to be synchronous?
```suggestion
    pub fn frame_display(&mut self) -> Result<ImageRaw<Gray8>, ()> {
        let raw_image: ImageRaw<Gray8>;
        raw_image = ImageRaw::<Gray8>::new(self.bytes.as_ref(), self.width);
        Ok(raw_image)
    }

    pub fn frame_clear(&mut self) {
```

Reply:
yes, that makes sense. Sorry, this should have been obvious to me :)